feat: u!-prefix support for user-defined types and function signatures (#154)
DataDog/substrait-explain: GRAMMAR.md

All types follow this general pattern:

```text
type := type_name anchor? urn_anchor? nullability? parameters?
type_name := "u!" identifier | identifier

> ```

Author: We may not need all types to support quoted names, therefore I think this can be ignored.

All types follow this general pattern:

```text
type := ("u!")? identifier anchor? urn_anchor? nullability? parameters?
```

Where:

> - **`identifier`** - The type name (case-insensitive, lowercase preferred), e.g. `geo_point` or `"my type"`. Both `u!json` and `json` refer to the same extension; the `u!` prefix is stripped at storage time.

Reviewer: 1. `"my type"` is a quoted name, we don't support that anymore
2. Let's note that `u!` is accepted but not recommended

```suggestion
- **`identifier`** - The type name (case-insensitive, lowercase preferred), e.g. `geo_point` or `my_type`. `u!my_type` syntax is accepted, but not recommended; the `u!` will be dropped - e.g. both `u!json` and `json` refer to the same extension.
```

All types follow this general pattern:

```text

> type := ("u!")? (identifier | quoted_name) anchor? urn_anchor? nullability? parameters?

Reviewer: ```suggestion
type := "u!"? identifier anchor? urn_anchor? nullability? parameters?
```

As above - I think we can drop quoted name support, as it isn't really supported here.

And anyway, let's try and be consistent within this doc: `name` is defined above as `identifier / quoted_name`, so no need to change that; and also, in EBNF the right symbol is `/`, not `|`. This doc is meant to be a clear EBNF grammar for the user; the syntax should match the Pest grammar, but the way they are described should fit this documentation, not simply repeat the Pest grammar.

```

Where:

> - **`identifier | quoted_name`** - The type name (case-insensitive, lowercase preferred), e.g. `geo_point` or `"my type"`. Both `u!json` and `json` refer to the same extension; the `u!` prefix is stripped at storage time.

Reviewer: Same, see above

#### Syntax

`udt_name anchor? urn_anchor? nullability? parameters?`

where `udt_name` is `"u!" (identifier | quoted_name) | identifier | quoted_name`.
#### Key differences from standard Substrait

- Adds optional `anchor` and `urn_anchor` for extension references

> - The `u!` prefix is part of the type name: a type declared as `u!json` must be referenced as `u!json`, not bare `json`

Reviewer: Let's reverse this, for ease of use:

```suggestion
- The `u!` prefix is not part of the type name: a type declared as `u!json` or as `json` can be referenced as `u!json` or as simply `json`.
```

#### Syntax

> `function_call := compound_function_signature anchor? urn_anchor? "(" (expression ("," expression)*)? ")" ":" type`

Reviewer: ```suggestion
`function_call := function_signature anchor? urn_anchor? "(" (expression ("," expression)*)? ")" ":" type`
```

Let's be consistent here in the naming

Reviewer: Can we apply this?
